# Importing pandas & Loading dataset

In [5]:
import pandas as pd

In [8]:
# Loading the csv file with pandas

In [12]:
df = pd.read_csv("~/Desktop/IP2/Amazon Sale Report.csv")

/var/folders/qn/xn_qh6qs56g4kyg5vcjpns8r0000gn/T/ipykernel_25368/3823463783.py:1: DtypeWarning: Columns (23) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("~/Desktop/IP2/Amazon Sale Report.csv")


# Exploring the dataset

In [20]:
# Checking DataTypes & general information
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 128975 entries, 0 to 128974
Data columns (total 24 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   index               128975 non-null  int64  
 1   Order ID            128975 non-null  object 
 2   Date                128975 non-null  object 
 3   Status              128975 non-null  object 
 4   Fulfilment          128975 non-null  object 
 5   Sales Channel       128975 non-null  object 
 6   ship-service-level  128975 non-null  object 
 7   Style               128975 non-null  object 
 8   SKU                 128975 non-null  object 
 9   Category            128975 non-null  object 
 10  Size                128975 non-null  object 
 11  ASIN                128975 non-null  object 
 12  Courier Status      122103 non-null  object 
 13  Qty                 128975 non-null  int64  
 14  currency            121180 non-null  object 
 15  Amount              121180 non-nul

In [24]:
# Display the first 5 rows of the DataFrame
df.head()

index             Order ID      Date                        Status  \
0      0  405-8078784-5731545  04-30-22                     Cancelled   
1      1  171-9198151-1101146  04-30-22  Shipped - Delivered to Buyer   
2      2  404-0687676-7273146  04-30-22                       Shipped   
3      3  403-9615377-8133951  04-30-22                     Cancelled   
4      4  407-1069790-7240320  04-30-22                       Shipped   

  Fulfilment Sales Channel  ship-service-level    Style              SKU  \
0   Merchant      Amazon.in           Standard   SET389   SET389-KR-NP-S   
1   Merchant      Amazon.in           Standard  JNE3781  JNE3781-KR-XXXL   
2     Amazon      Amazon.in          Expedited  JNE3371    JNE3371-KR-XL   
3   Merchant      Amazon.in           Standard    J0341       J0341-DR-L   
4     Amazon      Amazon.in          Expedited  JNE3671  JNE3671-TU-XXXL   

        Category  ... currency  Amount    ship-city   ship-state  \
0            Set  ...      INR  647.62       MUMBAI  MAHARASHTRA   
1          kurta  ...      INR  406.00    BENGALURU    KARNATAKA   
2          kurta  ...      INR  329.00  NAVI MUMBAI  MAHARASHTRA   
3  Western Dress  ...      INR  753.33   PUDUCHERRY   PUDUCHERRY   
4            Top  ...      INR  574.00      CHENNAI   TAMIL NADU   

  ship-postal-code  ship-country  \
0         400081.0            IN   
1         560085.0            IN   
2         410210.0            IN   
3         605008.0            IN   
4         600073.0            IN   

                                       promotion-ids    B2B  fulfilled-by  \
0                                                NaN  False     Easy Ship   
1  Amazon PLCC Free-Financing Universal Merchant ...  False     Easy Ship   
2       IN Core Free Shipping 2015/04/08 23-48-5-108   True           NaN   
3                                                NaN  False     Easy Ship   
4                                                NaN  False           NaN   

  Unnamed: 22  
0         NaN  
1         NaN  
2         NaN  
3         NaN  
4         NaN  

[5 rows x 24 columns]

In [27]:
# Dimension of the Dataset
df.shape

(128975, 24)

In [31]:
df.describe()

index            Qty         Amount  ship-postal-code
count  128975.000000  128975.000000  121180.000000     128942.000000
mean    64487.000000       0.904431     648.561465     463966.236509
std     37232.019822       0.313354     281.211687     191476.764941
min         0.000000       0.000000       0.000000     110001.000000
25%     32243.500000       1.000000     449.000000     382421.000000
50%     64487.000000       1.000000     605.000000     500033.000000
75%     96730.500000       1.000000     788.000000     600024.000000
max    128974.000000      15.000000    5584.000000     989898.000000

In [34]:
# Display the count of null values
df.isnull().sum()

index                     0
Order ID                  0
Date                      0
Status                    0
Fulfilment                0
Sales Channel             0
ship-service-level        0
Style                     0
SKU                       0
Category                  0
Size                      0
ASIN                      0
Courier Status         6872
Qty                       0
currency               7795
Amount                 7795
ship-city                33
ship-state               33
ship-postal-code         33
ship-country             33
promotion-ids         49153
B2B                       0
fulfilled-by          89698
Unnamed: 22           49050
dtype: int64

In [37]:
# Calculate the percentage of missing values for each column
missing_percentage = df.isnull().mean() * 100

# Display the percentage of missing values
print(missing_percentage)


index                  0.000000
Order ID               0.000000
Date                   0.000000
Status                 0.000000
Fulfilment             0.000000
Sales Channel          0.000000
ship-service-level     0.000000
Style                  0.000000
SKU                    0.000000
Category               0.000000
Size                   0.000000
ASIN                   0.000000
Courier Status         5.328164
Qty                    0.000000
currency               6.043807
Amount                 6.043807
ship-city              0.025586
ship-state             0.025586
ship-postal-code       0.025586
ship-country           0.025586
promotion-ids         38.110487
B2B                    0.000000
fulfilled-by          69.546811
Unnamed: 22           38.030626
dtype: float64


In [45]:
# Checking the count of unique values in the columns
df.nunique().to_frame(name='Count of unique values')

Count of unique values
index                               128975
Order ID                            120378
Date                                    91
Status                                  13
Fulfilment                               2
Sales Channel                            2
ship-service-level                       2
Style                                 1377
SKU                                   7195
Category                                 9
Size                                    11
ASIN                                  7190
Courier Status                           3
Qty                                     10
currency                                 1
Amount                                1410
ship-city                             8955
ship-state                              69
ship-postal-code                      9459
ship-country                             1
promotion-ids                         5787
B2B                                      2
fulfilled-by                             1
Unnamed: 22                              1

In [47]:
# Checking the unique values in the columns
df.apply(pd.unique).to_frame(name='Unique Values')

Unique Values
index               [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13,...
Order ID            [405-8078784-5731545, 171-9198151-1101146, 404...
Date                [04-30-22, 04-29-22, 04-28-22, 04-27-22, 04-26...
Status              [Cancelled, Shipped - Delivered to Buyer, Ship...
Fulfilment                                         [Merchant, Amazon]
Sales Channel                                 [Amazon.in, Non-Amazon]
ship-service-level                              [Standard, Expedited]
Style               [SET389, JNE3781, JNE3371, J0341, JNE3671, SET...
SKU                 [SET389-KR-NP-S, JNE3781-KR-XXXL, JNE3371-KR-X...
Category            [Set, kurta, Western Dress, Top, Ethnic Dress,...
Size                 [S, 3XL, XL, L, XXL, XS, 6XL, M, 4XL, 5XL, Free]
ASIN                [B09KXVBD7Z, B09K3WFS32, B07WV4JV4D, B099NRCT7...
Courier Status                   [nan, Shipped, Cancelled, Unshipped]
Qty                                  [0, 1, 2, 15, 3, 9, 13, 5, 4, 8]
currency                                                   [INR, nan]
Amount              [647.62, 406.0, 329.0, 753.33, 574.0, 824.0, 6...
ship-city           [MUMBAI, BENGALURU, NAVI MUMBAI, PUDUCHERRY, C...
ship-state          [MAHARASHTRA, KARNATAKA, PUDUCHERRY, TAMIL NAD...
ship-postal-code    [400081.0, 560085.0, 410210.0, 605008.0, 60007...
ship-country                                                [IN, nan]
promotion-ids       [nan, Amazon PLCC Free-Financing Universal Mer...
B2B                                                     [False, True]
fulfilled-by                                         [Easy Ship, nan]
Unnamed: 22                                              [nan, False]

# Deleting & Cleaning Columns 

### Upon review, it has been determined that analyzing the data at the city level would introduce excessive granularity. 
### Hence, focusing the analysis on the state level is deemed more appropriate for this holistic sales report for India.

In [59]:
# Drop columns that are deemed unnecessary for the state-level analysis
df.drop(columns=["ASIN", "SKU", "index", "Style", "Order ID", "ship-city", "Unnamed: 22", "fulfilled-by", "promotion-ids"], inplace=True)

In [62]:
# Convert values in 'ship-state' column to title case
df["ship-state"] = df["ship-state"].str.title()
df["ship-state"]

0          Maharashtra
1            Karnataka
2          Maharashtra
3           Puducherry
4           Tamil Nadu
              ...     
128970       Telangana
128971         Haryana
128972       Telangana
128973         Gujarat
128974    Chhattisgarh
Name: ship-state, Length: 128975, dtype: object

In [64]:
# Display the count of null values
df.isnull().sum()

Date                     0
Status                   0
Fulfilment               0
Sales Channel            0
ship-service-level       0
Category                 0
Size                     0
Courier Status        6872
Qty                      0
currency              7795
Amount                7795
ship-state              33
ship-postal-code        33
ship-country            33
B2B                      0
dtype: int64

### Given the occurrence of missing values in the Amount column, it is essential to determine whether these gaps are primarily due to canceled orders or if they stem from other contributing factors.

In [71]:
# Filter the rows where 'Amount' is NaN
nan_amount_rows = df[df['Amount'].isna()]

# Display the unique values of 'Status' and 'Courier Status' in these rows
nan_amount_status = nan_amount_rows['Status'].unique()
nan_amount_courier_status = nan_amount_rows['Courier Status'].unique()

# Count unique combinations of 'Status' and 'Courier Status'
nan_amount_rows_summary = nan_amount_rows[['Status', 'Courier Status']].value_counts()

# Display the unique statuses and courier statuses, and the count of these combinations
print("Unique 'Status' values where 'Amount' is NaN:")
print(nan_amount_status)

print("\nUnique 'Courier Status' values where 'Amount' is NaN:")
print(nan_amount_courier_status)

print("\nCounts of 'Status' and 'Courier Status' combinations where 'Amount' is NaN:")
print(nan_amount_rows_summary)

Unique 'Status' values where 'Amount' is NaN:
['Cancelled' 'Shipped' 'Shipped - Delivered to Buyer' 'Pending'
 'Shipped - Returned to Seller' 'Shipping']

Unique 'Courier Status' values where 'Amount' is NaN:
['Cancelled' nan 'Unshipped']

Counts of 'Status' and 'Courier Status' combinations where 'Amount' is NaN:
Status     Courier Status
Cancelled  Cancelled         5840
Shipped    Unshipped          115
           Cancelled           93
Shipping   Unshipped            8
Pending    Cancelled            2
Cancelled  Unshipped            1
Name: count, dtype: int64


### It has been observed that there is necessary to further investigate the rows where the "Courier Status" is Nan

In [76]:
# Filter rows where both 'Amount' and 'Courier Status' are NaN
nan_amount_nan_courier_status_rows = df[df['Amount'].isna() & df['Courier Status'].isna()]

# Display relevant information
print("Rows with both 'Amount' and 'Courier Status' missing:")
print(nan_amount_nan_courier_status_rows[['Courier Status', 'ship-state']])

Rows with both 'Amount' and 'Courier Status' missing:
       Courier Status     ship-state
29                NaN          Assam
250               NaN      Karnataka
563               NaN  Uttar Pradesh
672               NaN        Haryana
818               NaN      Karnataka
...               ...            ...
128277            NaN      Karnataka
128560            NaN     Tamil Nadu
128571            NaN  Uttar Pradesh
128808            NaN      Karnataka
128855            NaN         Kerala

[1736 rows x 2 columns]


In [80]:
# Filter rows where 'Courier Status' is NaN
nan_courier_status_rows = df[df['Courier Status'].isna()]

# Check if these rows have NaN in 'Amount' or if they have non-NaN values
nan_courier_status_amount_summary = nan_courier_status_rows['Amount'].isna().value_counts()

# Display the summary
print("Summary of 'Amount' where 'Courier Status' is NaN:")
print(nan_courier_status_amount_summary)

Summary of 'Amount' where 'Courier Status' is NaN:
Amount
False    5136
True     1736
Name: count, dtype: int64


In [83]:
# Additional exploration to understand the distribution of missing values
missing_amount_courier_status_summary = nan_amount_nan_courier_status_rows['ship-state'].value_counts()

print("Distribution of missing 'Amount' and 'Courier Status' by 'ship-state':")
print(missing_amount_courier_status_summary)

Distribution of missing 'Amount' and 'Courier Status' by 'ship-state':
ship-state
Maharashtra           239
Karnataka             232
Uttar Pradesh         189
Telangana             156
Tamil Nadu            134
Kerala                108
Delhi                  91
West Bengal            85
Andhra Pradesh         82
Gujarat                53
Madhya Pradesh         47
Rajasthan              43
Haryana                37
Assam                  35
Jharkhand              27
Punjab                 27
Uttarakhand            24
Bihar                  23
Himachal Pradesh       18
Odisha                 18
Jammu & Kashmir        14
Goa                    13
Chhattisgarh           11
Tripura                 5
Dadra And Nagar         4
Arunachal Pradesh       3
Manipur                 3
Sikkim                  3
Mizoram                 2
Meghalaya               2
Chandigarh              2
Andaman & Nicobar       2
Puducherry              2
New Delhi               1
Nagaland                1
Name: co

In [86]:
# Analyze Distribution of 'Status' for Rows with Missing 'Amount' and 'Courier Status'
missing_amount_courier_status_status_summary = nan_amount_nan_courier_status_rows['Status'].value_counts()

print("Distribution of missing 'Amount' and 'Courier Status' by 'Status':")
print(missing_amount_courier_status_status_summary)

Distribution of missing 'Amount' and 'Courier Status' by 'Status':
Status
Cancelled                       1725
Shipped - Delivered to Buyer       8
Shipped - Returned to Seller       3
Name: count, dtype: int64


In [91]:
# Filtering on the 8 suspect rows where "Status" is "Shipped - Delivered to Buyer"
filtered_df = df[(df['Status'] == 'Shipped - Delivered to Buyer') &
                 (df['Qty'] == 0.0) &
                 (df['Amount'] == 0.0) &
                 (pd.isna(df['Courier Status']))]

# Display the result
print(filtered_df)


Empty DataFrame
Columns: [Date, Status, Fulfilment, Sales Channel , ship-service-level, Category, Size, Courier Status, Qty, currency, Amount, ship-state, ship-postal-code, ship-country, B2B]
Index: []


#### For some reason is my Dataframe empty 

-------------------------------------------------------------------------------------------------------------------

# Based on these rows, we can draw several important conclusions regarding the inconsistencies in the data:

#### All the rows have "Shipped - Delivered to Buyer" as status, meaning that the items were reportedly delivered to the customer. However, the "Courier Status" is marked as NaN. This raises a red flag because the courier status should provide information on the shipment progress. The absence of this information indicates that there is either a gap in the tracking system or incomplete data entry.

#### For all the rows, the quantity is listed as 0, and the amount is also 0. This is unusual, as it tells us no products were sold or shipped according to these columns, despite the "Shipped - Delivered to Buyer" status. This could indicate a data error or inconsistency in the order management system.

#### From a business perspective, these rows need further investigation to ensure accuracy. If the records represent actual sales, then the missing data (amount, quantity, courier status) could lead to incorrect financial reporting. Alternatively, if these are invalid or canceled orders, they should be cleaned from the dataset to prevent skewed analysis.


------------------------------------------------------------------------------------------------------------------------------------------------

In [113]:
df["Status"].unique()

array(['Cancelled', 'Shipped - Delivered to Buyer', 'Shipped',
       'Shipped - Returned to Seller', 'Shipped - Rejected by Buyer',
       'Shipped - Lost in Transit', 'Shipped - Out for Delivery',
       'Shipped - Returning to Seller', 'Shipped - Picked Up', 'Pending',
       'Pending - Waiting for Pick Up', 'Shipped - Damaged', 'Shipping'],
      dtype=object)

### Standardizing "Status" Values and Renaming "Qty" Column for Clarity

In [117]:
# Replacing values in the 'Status' column
df['Status'] = df['Status'].replace({
    'Shipping': 'Shipped',
    'Shipped - Returning to Seller': 'Shipped - Returned to Seller'
})

# Rename the 'Qty' column to 'Quantity'
df = df.rename(columns={
    'Qty': 'Quantity'
})

In [ ]:
# Verifying the Renaming of "Qty" Column
df.head()

In [123]:
# Verifying the standardizing of values in the "Status" Column
df["Status"].unique()

array(['Cancelled', 'Shipped - Delivered to Buyer', 'Shipped',
       'Shipped - Returned to Seller', 'Shipped - Rejected by Buyer',
       'Shipped - Lost in Transit', 'Shipped - Out for Delivery',
       'Shipped - Picked Up', 'Pending', 'Pending - Waiting for Pick Up',
       'Shipped - Damaged'], dtype=object)

In [126]:
# Replace NaN values with 'INR' in the 'currency' column
df['currency'].fillna('INR', inplace=True)

# Check the unique values to confirm the replacement
df['currency'].unique()

array(['INR'], dtype=object)

In [133]:
df.head()

Date                        Status Fulfilment Sales Channel   \
0  04-30-22                     Cancelled   Merchant      Amazon.in   
1  04-30-22  Shipped - Delivered to Buyer   Merchant      Amazon.in   
2  04-30-22                       Shipped     Amazon      Amazon.in   
3  04-30-22                     Cancelled   Merchant      Amazon.in   
4  04-30-22                       Shipped     Amazon      Amazon.in   

  ship-service-level       Category Size Courier Status  Quantity currency  \
0           Standard            Set    S            NaN         0      INR   
1           Standard          kurta  3XL        Shipped         1      INR   
2          Expedited          kurta   XL        Shipped         1      INR   
3           Standard  Western Dress    L            NaN         0      INR   
4          Expedited            Top  3XL        Shipped         1      INR   

   Amount   ship-state  ship-postal-code ship-country    B2B  
0  647.62  Maharashtra          400081.0           IN  False  
1  406.00    Karnataka          560085.0           IN  False  
2  329.00  Maharashtra          410210.0           IN   True  
3  753.33   Puducherry          605008.0           IN  False  
4  574.00   Tamil Nadu          600073.0           IN  False

# Observed in row 3 that "Status" is marked as "Cancelled," "Courier Status" is missing (NaN), "Quantity" is 0, but "Amount" is recorded as 753.33
## These amounts may distort the data. It's unclear whether the payment has been received, has already been refunded, or is pending a refund.

## Its decided that when 
#### - Status = "Cancelled"
#### - Courier Status is NaN, 
#### - Quantity = 0
#### - Amount > 0

### These data entries will be excluded from the analysis. This decision is based on the inability to confidently verify whether these amounts represent actual revenue. 
### I would suggest conducting a root cause analysis in a real-life scenario based on these specific inconsistencies

In [161]:
# Creating a variable in order to varify if the loop was succesfull
before_update_amount = df[(df['Status'] == 'Cancelled') &
                                   (df['Courier Status'].isna()) &
                                   (df['Quantity'] == 0) &
                                   (df['Amount'] > 0)].shape[0]

In [155]:
before_update_amount

5136

In [157]:
# Loop to replace the amount based on these 4 condition 
for index, row in df.iterrows():
    if (row['Status'] == 'Cancelled' and
        pd.isna(row['Courier Status']) and
        row['Quantity'] == 0 and
        row['Amount'] > 0):
        df.at[index, 'Amount'] = 0

In [163]:
# Verifying that the loop worked
after_update_amount = df[(df['Status'] == 'Cancelled') &
                                  (df['Courier Status'].isna()) &
                                  (df['Quantity'] == 0) &
                                  (df['Amount'] > 0)].shape[0]
after_update_amount

0

In [166]:
df.head()

Date                        Status Fulfilment Sales Channel   \
0  04-30-22                     Cancelled   Merchant      Amazon.in   
1  04-30-22  Shipped - Delivered to Buyer   Merchant      Amazon.in   
2  04-30-22                       Shipped     Amazon      Amazon.in   
3  04-30-22                     Cancelled   Merchant      Amazon.in   
4  04-30-22                       Shipped     Amazon      Amazon.in   

  ship-service-level       Category Size Courier Status  Quantity currency  \
0           Standard            Set    S            NaN         0      INR   
1           Standard          kurta  3XL        Shipped         1      INR   
2          Expedited          kurta   XL        Shipped         1      INR   
3           Standard  Western Dress    L            NaN         0      INR   
4          Expedited            Top  3XL        Shipped         1      INR   

   Amount   ship-state  ship-postal-code ship-country    B2B  
0     0.0  Maharashtra          400081.0           IN  False  
1   406.0    Karnataka          560085.0           IN  False  
2   329.0  Maharashtra          410210.0           IN   True  
3     0.0   Puducherry          605008.0           IN  False  
4   574.0   Tamil Nadu          600073.0           IN  False

In [169]:
df.isnull().sum()

Date                     0
Status                   0
Fulfilment               0
Sales Channel            0
ship-service-level       0
Category                 0
Size                     0
Courier Status        6872
Quantity                 0
currency                 0
Amount                7795
ship-state              33
ship-postal-code        33
ship-country            33
B2B                      0
dtype: int64

In [171]:
# Change column names to title case
df.columns = df.columns.str.title()

# Replace NaN values
df['Amount'].fillna(0, inplace=True)
df['Courier Status'].fillna('Unknown', inplace=True)

# Verify changes
df.isnull().sum()
df.head()


Date                        Status Fulfilment Sales Channel   \
0  04-30-22                     Cancelled   Merchant      Amazon.in   
1  04-30-22  Shipped - Delivered to Buyer   Merchant      Amazon.in   
2  04-30-22                       Shipped     Amazon      Amazon.in   
3  04-30-22                     Cancelled   Merchant      Amazon.in   
4  04-30-22                       Shipped     Amazon      Amazon.in   

  Ship-Service-Level       Category Size Courier Status  Quantity Currency  \
0           Standard            Set    S        Unknown         0      INR   
1           Standard          kurta  3XL        Shipped         1      INR   
2          Expedited          kurta   XL        Shipped         1      INR   
3           Standard  Western Dress    L        Unknown         0      INR   
4          Expedited            Top  3XL        Shipped         1      INR   

   Amount   Ship-State  Ship-Postal-Code Ship-Country    B2B  
0     0.0  Maharashtra          400081.0           IN  False  
1   406.0    Karnataka          560085.0           IN  False  
2   329.0  Maharashtra          410210.0           IN   True  
3     0.0   Puducherry          605008.0           IN  False  
4   574.0   Tamil Nadu          600073.0           IN  False

In [175]:
# Verify changes
df.isnull().sum()

Date                   0
Status                 0
Fulfilment             0
Sales Channel          0
Ship-Service-Level     0
Category               0
Size                   0
Courier Status         0
Quantity               0
Currency               0
Amount                 0
Ship-State            33
Ship-Postal-Code      33
Ship-Country          33
B2B                    0
dtype: int64

##### Replace NaN values in 'Ship-Country' with 'IN'

In [177]:
df['Ship-Country'].fillna('IN', inplace=True)

In [181]:
# Verify changes
df.isnull().sum()

Date                   0
Status                 0
Fulfilment             0
Sales Channel          0
Ship-Service-Level     0
Category               0
Size                   0
Courier Status         0
Quantity               0
Currency               0
Amount                 0
Ship-State            33
Ship-Postal-Code      33
Ship-Country           0
B2B                    0
dtype: int64

In [187]:
# Convert values in 'Category' column to title case
df["Category"] = df["Category"].str.title()
df["Category"]

0                   Set
1                 Kurta
2                 Kurta
3         Western Dress
4                   Top
              ...      
128970            Kurta
128971              Set
128972    Western Dress
128973              Set
128974              Set
Name: Category, Length: 128975, dtype: object

### Standardizing 'Ship-State'

In [196]:
# Display unique values in the 'Ship-State' column
df['Ship-State'].unique()



array(['Maharashtra', 'Karnataka', 'Puducherry', 'Tamil Nadu',
       'Uttar Pradesh', 'Chandigarh', 'Telangana', 'Andhra Pradesh',
       'Rajasthan', 'Delhi', 'Haryana', 'Assam', 'Jharkhand',
       'Chhattisgarh', 'Odisha', 'Kerala', 'Madhya Pradesh',
       'West Bengal', 'Nagaland', 'Gujarat', 'Uttarakhand', 'Bihar',
       'Jammu & Kashmir', 'Punjab', 'Himachal Pradesh',
       'Arunachal Pradesh', 'Manipur', 'Goa', 'Meghalaya', 'Tripura',
       'Ladakh', 'Dadra And Nagar', 'Sikkim', nan, 'Andaman & Nicobar ',
       'Rajshthan', 'Nl', 'Mizoram', 'New Delhi',
       'Punjab/Mohali/Zirakpur', 'Rj', 'Orissa', 'Lakshadweep', 'Pb',
       'Apo', 'Ar', 'Pondicherry', 'Rajsthan'], dtype=object)

In [202]:
# list of valid states
valid_states = [
    'Maharashtra', 'Karnataka', 'Puducherry', 'Tamil Nadu', 'Uttar Pradesh',
    'Chandigarh', 'Telangana', 'Andhra Pradesh', 'Rajasthan', 'Delhi', 'Haryana',
    'Assam', 'Jharkhand', 'Chhattisgarh', 'Odisha', 'Kerala', 'Madhya Pradesh',
    'West Bengal', 'Nagaland', 'Gujarat', 'Uttarakhand', 'Bihar', 'Jammu & Kashmir',
    'Punjab', 'Himachal Pradesh', 'Arunachal Pradesh', 'Manipur', 'Goa', 'Meghalaya',
    'Tripura', 'Ladakh', 'Dadra And Nagar', 'Sikkim', 'Andaman & Nicobar', 'Mizoram',
    'New Delhi', 'Lakshadweep'
]

# dictionary for corrections
corrections = {
    'Rajshthan': 'Rajasthan',
    'Rajsthan': 'Rajasthan',
    'Nl': 'Nagaland',
    'Orissa': 'Odisha',
    'Pondicherry': 'Puducherry',
    'Pb': 'Punjab',
    'Apo': 'Andaman & Nicobar',
    'Ar': 'Arunachal Pradesh',
    'Rj': 'Rajasthan',
    'Rajsthan': 'Rajasthan',
    'Punjab/Mohali/Zirakpur': 'Punjab'
}

# Replace incorrect names with correct ones
df['Ship-State'] = df['Ship-State'].replace(corrections)

# Remove or replace invalid entries
df['Ship-State'] = df['Ship-State'].apply(lambda x: x if x in valid_states else pd.NA)

# Check the cleaned values
print(df['Ship-State'].unique())


['Maharashtra' 'Karnataka' 'Puducherry' 'Tamil Nadu' 'Uttar Pradesh'
 'Chandigarh' 'Telangana' 'Andhra Pradesh' 'Rajasthan' 'Delhi' 'Haryana'
 'Assam' 'Jharkhand' 'Chhattisgarh' 'Odisha' 'Kerala' 'Madhya Pradesh'
 'West Bengal' 'Nagaland' 'Gujarat' 'Uttarakhand' 'Bihar'
 'Jammu & Kashmir' 'Punjab' 'Himachal Pradesh' 'Arunachal Pradesh'
 'Manipur' 'Goa' 'Meghalaya' 'Tripura' 'Ladakh' 'Dadra And Nagar' 'Sikkim'
 <NA> 'Mizoram' 'New Delhi' 'Lakshadweep' 'Andaman & Nicobar']


In [205]:
# Extracting DataFrame to a CSV file
df.to_csv('Amzn_sales_data.csv', index=False)